sdk/identity/azure-identity/src/client_certificate_credential.cpp

```diff
@@ -0,0 +1,296 @@
+          }
+        }
+
+        BIO_free(bio);
+        bio = nullptr;
+
+        // Get certificate thumbprint:
+        {
+          using Azure::Core::_internal::Base64Url;
+
+          std::string thumbprintHexStr;
+          std::string thumbprintBase64Str;
+          {
+            std::vector<unsigned char> mdVec(EVP_MAX_MD_SIZE);
+            {
+              unsigned int mdLen = 0;
+              const auto digestResult = X509_digest(x509, EVP_sha1(), mdVec.data(), &mdLen);
+
+              X509_free(x509);
+              x509 = nullptr;
+
+              if (!digestResult)
+              {
+                throw AuthenticationException("Failed to get certificate thumbprint.");
+              }
+
+              // Drop unused buffer space:
+              const auto mdLenSz = static_cast<decltype(mdVec)::size_type>(mdLen);
+              if (mdVec.size() > mdLenSz)
+              {
+                mdVec.resize(mdLenSz);
+              }
+
+              // Get thumbprint as hex string:
+              {
+                std::ostringstream thumbprintStream;
+                for (const auto md : mdVec)
+                {
+                  thumbprintStream << std::uppercase << std::hex << std::setfill('0')
+                                   << std::setw(2) << static_cast<int>(md);
+                }
+                thumbprintHexStr = thumbprintStream.str();
+              }
+            }
+
+            // Get thumbprint as Base64:
+            thumbprintBase64Str = Base64Url::Base64UrlEncode(ToUInt8Vector(mdVec));
+          }
+
+          // Form a JWT token:
+          const auto tokenHeader = std::string("{\"x5t\":\"") + thumbprintBase64Str
+              + "\",\"kid\":\"" + thumbprintHexStr + "\",\"alg\":\"RS256\",\"typ\":\"JWT\"}";
+
+          const auto tokenHeaderVec
+              = std::vector<std::string::value_type>(tokenHeader.begin(), tokenHeader.end());
+
+          m_tokenHeaderEncoded = Base64Url::Base64UrlEncode(ToUInt8Vector(tokenHeaderVec));
+        }
+      }
+      else
+      {
+        throw AuthenticationException("Failed to open certificate file.");
+      }
+    }
+
+    using Azure::Core::Url;
+    {
+
+      m_requestUrl = Url("https://login.microsoftonline.com/");
+      m_requestUrl.AppendPath(tenantId);
+      m_requestUrl.AppendPath("oauth2/v2.0/token");
+    }
+
+    m_tokenPayloadStaticPart = std::string("{\"aud\":\"") + m_requestUrl.GetAbsoluteUrl()
+        + "\",\"iss\":\"" + clientId + "\",\"sub\":\"" + clientId + "\",\"jti\":\"";
+
+    {
+      std::ostringstream body;
+      body
+          << "grant_type=client_credentials"
+             "&client_assertion_type="
+             "urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer" // cspell:disable-line
+             "&client_id="
+          << Url::Encode(clientId);
+
+      m_requestBody = body.str();
+    }
+  }
+  catch (...)
+  {
+    if (bio != nullptr)
+    {
+      BIO_free(bio);
+    }
```

Comment:
I was thinking about it, but moving it to another place, and declaring, does add extra lines, so in addition to benefits brings disadvantages, so I think it is not yet justify creation of an extra entity. If anything, I would rather make a generic RAII class that you initialize with openssl resource pointer, and parameterize with `free` function. But IMO, since everything happens inside one method, it is not yet justified, but it is on a borderline. It is also that we need to `release()` one of the pointers (we only call `free()` in case of error inside constructor), so I imagine the class will be 10-20 lines, which is comparable to the current code that handles it. Plus, you'd need to parameterize it with a free function pointer, and another problem is that one of the free() functions returns int, others do return void. It will take more code to write a template for that, than the regular approach. Plus, the template code won't be as easy to read. Or, alternatively, there needs to be function<T> parameter, but it is one more indirect call.
Other thing, if we implement something with GetRequestBody(), its constructor will basically need to have the exact same constructor body, as we're not gaining anything.